compiler/rustc_typeck/src/check/generator_interior/drop_ranges/record_consumed_borrow.rs

```diff
@@ -0,0 +1,127 @@
+use crate::{
+    check::FnCtxt,
+    expr_use_visitor::{self, ExprUseVisitor},
+};
+use hir::{def_id::DefId, Body, HirId, HirIdMap, HirIdSet};
+use rustc_hir as hir;
+use rustc_middle::hir::{
+    map::Map,
+    place::{Place, PlaceBase},
+};
+use rustc_middle::ty;
+
+pub fn find_consumed_and_borrowed<'a, 'tcx>(
+    fcx: &'a FnCtxt<'a, 'tcx>,
+    def_id: DefId,
+    body: &'tcx Body<'tcx>,
+) -> ConsumedAndBorrowedPlaces {
+    let mut expr_use_visitor = ExprUseDelegate::new(fcx.tcx.hir());
+    expr_use_visitor.consume_body(fcx, def_id, body);
+    expr_use_visitor.places
+}
+
+pub struct ConsumedAndBorrowedPlaces {
+    /// Records the variables/expressions that are dropped by a given expression.
+    ///
+    /// The key is the hir-id of the expression, and the value is a set or hir-ids for variables
+    /// or values that are consumed by that expression.
+    ///
+    /// Note that this set excludes "partial drops" -- for example, a statement like `drop(x.y)` is
+    /// not considered a drop of `x`, although it would be a drop of `x.y`.
```

Comment:
Does it hurt us to track drops of `x.y`? If someone had something like
```rust
struct Thing {
    x: Sendable,
    y: NonSend
}

let thing = Thing { ... };

foo().await;
```
then they'd get an error because of the `NonSend` field on `Thing`. This lets them write `drop(thing.y)` before the `await` to work around the error.

Currently we track everything that `ExprUseVisitor` tells us is consumed, which seems nice from a standpoint of being consistent with other code.

Reply:
Something feels funny to me about this. I think I misunderstood what was going on. I expected we were tracking drops of *temporaries* but it sounds like we are identifying drops of partial paths via a HirId? That seems a bit sketchy to me. I'll have to re-read the code with this in mind.

Reply:
A lot of the sketchiness should be contained within [`for_each_consumable`](https://github.com/rust-lang/rust/pull/91032/files#diff-ae293363720f752a3d06e276a8c0090f06370003699f8ff0885eb1fe39ac134aR56) and [`reinit_expr`](https://github.com/rust-lang/rust/pull/91032/files#diff-1bcc2a45ce477888aa0469a2a536779dafd322470950ee45f3e401a8512aca59R95), so that's where I'd focus on in the reread.